File: vito/core/src/test/java/com/facebook/fresco/vito/core/impl/HierarcherImplTest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.core.impl;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.AdditionalMatchers.not;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.PointF;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.util.DisplayMetrics;
import com.facebook.drawee.drawable.RoundedBitmapDrawable;
import com.facebook.drawee.drawable.RoundedColorDrawable;
import com.facebook.drawee.drawable.ScaleTypeDrawable;
import com.facebook.drawee.drawable.ScalingUtils;
import com.facebook.fresco.vito.core.Hierarcher;
import com.facebook.fresco.vito.options.ImageOptions;
import com.facebook.fresco.vito.options.ImageOptionsDrawableFactory;
import com.facebook.fresco.vito.options.RoundingOptions;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

@RunWith(RobolectricTestRunner.class)
public class HierarcherImplTest {

  private static final int RES_ID = 123;
  private static final int INVALID_RES_ID = 999;

  private Resources mResources;
  private DisplayMetrics mDisplayMetrics;
  private Drawable mDrawable;

  private Hierarcher mHierarcher;

  private ImageOptionsDrawableFactory mDrawableFactory;

  @Before
  public void setup() {
    mResources = mock(Resources.class);
    mDrawable = mock(Drawable.class);
    mDrawableFactory = mock(ImageOptionsDrawableFactory.class);
    mDisplayMetrics = new DisplayMetrics();
    when(mResources.getDisplayMetrics()).thenReturn(mDisplayMetrics);
    when(mResources.getDrawable(eq(RES_ID))).thenReturn(mDrawable);
    when(mResources.getDrawable(not(eq(RES_ID)))).thenThrow(new Resources.NotFoundException());

    mHierarcher = new HierarcherImpl(mDrawableFactory);
  }

  @Test
  public void testBuildPlaceholderRes() {
    ImageOptions options = ImageOptions.create().placeholderRes(RES_ID).build();

    Drawable errorDrawable = mHierarcher.buildPlaceholderDrawable(mResources, options);

    assertThat(errorDrawable).isNotNull();
    assertThat(errorDrawable).isInstanceOf(ScaleTypeDrawable.class);

    ScaleTypeDrawable scaleTypeDrawable = (ScaleTypeDrawable) errorDrawable;
    assertThat(scaleTypeDrawable.getScaleType())
        .isEqualTo(ImageOptions.defaults().getPlaceholderScaleType());
    assertThat(scaleTypeDrawable.getFocusPoint())
        .isEqualTo(ImageOptions.defaults().getPlaceholderFocusPoint());
    assertThat(scaleTypeDrawable.getCurrent()).isEqualTo(mDrawable);
  }

  @Test
  public void testBuildPlaceholderDrawable() {
    final Drawable expected = new ColorDrawable(Color.YELLOW);
    ImageOptions options =
        ImageOptions.create().placeholder(expected).placeholderScaleType(null).build();

    final Drawable result = mHierarcher.buildPlaceholderDrawable(mResources, options);

    assertThat(result).isEqualTo(expected);
  }

  @Test
  public void testBuildPlaceholderDrawableScale() {
    final Drawable expected = new ColorDrawable(Color.YELLOW);
    ImageOptions options =
        ImageOptions.create()
            .placeholder(expected)
            .placeholderScaleType(ScalingUtils.ScaleType.CENTER)
            .build();

    final Drawable result = mHierarcher.buildPlaceholderDrawable(mResources, options);

    assertThat(result).isExactlyInstanceOf(ScaleTypeDrawable.class);
    assertThat(result.getCurrent()).isEqualTo(expected);
  }

  @Test
  public void testApplyRoundingOptions_whenRoundAsCircle_thenReturnDrawable() {
    final Drawable drawable = new ColorDrawable(Color.YELLOW);
    when(mResources.getDrawable(RES_ID)).thenReturn(drawable);

    ImageOptions options = mock(ImageOptions.class);
    when(options.getRoundingOptions()).thenReturn(RoundingOptions.asCircle());
    when(options.getPlaceholderDrawable()).thenReturn(drawable);
    when(options.getPlaceholderApplyRoundingOptions()).thenReturn(true);

    Drawable result = mHierarcher.buildPlaceholderDrawable(mResources, options);

    assertThat(result).isExactlyInstanceOf(RoundedColorDrawable.class);

    when(options.getPlaceholderDrawable()).thenReturn(null);
    when(options.getPlaceholderRes()).thenReturn(RES_ID);

    result = mHierarcher.buildPlaceholderDrawable(mResources, options);

    assertThat(result).isExactlyInstanceOf(RoundedColorDrawable.class);
  }

  @Test
  public void testApplyRoundingOptions_whenRoundWithCornerRadius_thenReturnDrawable() {
    final BitmapDrawable drawable = mock(BitmapDrawable.class);
    final Bitmap bitmap = mock(Bitmap.class);
    when(drawable.getBitmap()).thenReturn(bitmap);

    ImageOptions options = mock(ImageOptions.class);
    when(options.getRoundingOptions()).thenReturn(RoundingOptions.forCornerRadiusPx(123));
    when(options.getPlaceholderDrawable()).thenReturn(drawable);
    when(options.getPlaceholderApplyRoundingOptions()).thenReturn(true);

    Drawable result = mHierarcher.buildPlaceholderDrawable(mResources, options);

    assertThat(result).isNotNull();
    assertThat(result).isInstanceOf(RoundedBitmapDrawable.class);
    assertThat(((RoundedBitmapDrawable) result).getRadii())
        .isEqualTo(new float[] {123, 123, 123, 123, 123, 123, 123, 123});
  }

  @Test
  public void testApplyRoundingOptions_whenRoundWithCornerRadii_thenReturnDrawable() {
    final BitmapDrawable drawable = mock(BitmapDrawable.class);
    final Bitmap bitmap = mock(Bitmap.class);
    when(drawable.getBitmap()).thenReturn(bitmap);

    ImageOptions options = mock(ImageOptions.class);
    when(options.getRoundingOptions()).thenReturn(RoundingOptions.forCornerRadii(1, 2, 3, 4));
    when(options.getPlaceholderDrawable()).thenReturn(drawable);
    when(options.getPlaceholderApplyRoundingOptions()).thenReturn(true);

    Drawable result = mHierarcher.buildPlaceholderDrawable(mResources, options);

    assertThat(result).isNotNull();
    assertThat(result).isInstanceOf(RoundedBitmapDrawable.class);
    assertThat(((RoundedBitmapDrawable) result).getRadii())
        .isEqualTo(new float[] {1, 1, 2, 2, 3, 3, 4, 4});
  }

  @Test
  public void testBuildErrorDrawable_whenNoScaleTypeSet_thenUseDefaultScaleType() {
    ImageOptions options = ImageOptions.create().errorRes(RES_ID).build();

    Drawable errorDrawable = mHierarcher.buildErrorDrawable(mResources, options);

    assertThat(errorDrawable).isNotNull();
    assertThat(errorDrawable).isInstanceOf(ScaleTypeDrawable.class);

    ScaleTypeDrawable scaleTypeDrawable = (ScaleTypeDrawable) errorDrawable;
    assertThat(scaleTypeDrawable.getScaleType())
        .isEqualTo(ImageOptions.defaults().getErrorScaleType());
    assertThat(scaleTypeDrawable.getFocusPoint())
        .isEqualTo(ImageOptions.defaults().getErrorFocusPoint());
    assertThat(scaleTypeDrawable.getCurrent()).isEqualTo(mDrawable);
  }

  @Test
  public void testBuildErrorDrawable_whenScaleTypeNull_thenDoNotWrapDrawable() {
    ImageOptions options = ImageOptions.create().errorRes(RES_ID).errorScaleType(null).build();

    Drawable errorDrawable = mHierarcher.buildErrorDrawable(mResources, options);

    assertThat(errorDrawable).isEqualTo(mDrawable);
  }

  @Test(expected = Resources.NotFoundException.class)
  public void testBuildErrorDrawable_whenInvalidResId_thenThrowNotFoundException() {
    ImageOptions options = ImageOptions.create().errorRes(INVALID_RES_ID).build();

    mHierarcher.buildErrorDrawable(mResources, options);
  }

  @Test
  public void testBuildErrorDrawable_whenScaleTypeSet_thenReturnScaleTypeDrawable() {
    PointF focusPoint = new PointF(100, 234);
    ImageOptions options =
        ImageOptions.create()
            .errorRes(RES_ID)
            .errorScaleType(ScalingUtils.ScaleType.FOCUS_CROP)
            .errorFocusPoint(focusPoint)
            .build();

    Drawable errorDrawable = mHierarcher.buildErrorDrawable(mResources, options);

    assertThat(errorDrawable).isNotNull();
    assertThat(errorDrawable).isInstanceOf(ScaleTypeDrawable.class);

    ScaleTypeDrawable scaleTypeDrawable = (ScaleTypeDrawable) errorDrawable;
    assertThat(scaleTypeDrawable.getScaleType()).isEqualTo(ScalingUtils.ScaleType.FOCUS_CROP);
    assertThat(scaleTypeDrawable.getFocusPoint()).isEqualTo(focusPoint);
    assertThat(scaleTypeDrawable.getCurrent()).isEqualTo(mDrawable);
  }

  @Test
  public void testBuildErrorDrawable_whenNotSet_thenReturnNopDrawable() {
    ImageOptions options = ImageOptions.create().build();

    Drawable errorDrawable = mHierarcher.buildErrorDrawable(mResources, options);

    assertThat(errorDrawable).isNull();
  }

  @Test
  public void testBuildProgressDrawable() {
    final Drawable drawable = new ColorDrawable(0x0);
    final ImageOptions imageOptions =
        ImageOptions.create()
            .progress(drawable)
            .progressScaleType(ScalingUtils.ScaleType.FIT_CENTER)
            .build();

    final Drawable actual = mHierarcher.buildProgressDrawable(mResources, imageOptions);
    assertThat(actual).isInstanceOf(ScaleTypeDrawable.class);
    final ScaleTypeDrawable scaleTypeActual = (ScaleTypeDrawable) actual;
    assertThat(scaleTypeActual.getScaleType()).isEqualTo(ScalingUtils.ScaleType.FIT_CENTER);
    assertThat(scaleTypeActual.getCurrent()).isEqualTo(drawable);
  }

  @Test
  public void testBuildActualImageWrapper() {
    final PointF expectedFocusPoint = new PointF(1, 2);
    final ImageOptions imageOptions =
        ImageOptions.create()
            .scale(ScalingUtils.ScaleType.FIT_CENTER)
            .focusPoint(expectedFocusPoint)
            .build();

    final Drawable actual = mHierarcher.buildActualImageWrapper(imageOptions, null);
    assertThat(actual).isInstanceOf(ScaleTypeDrawable.class);
    final ScaleTypeDrawable scaleTypeActual = (ScaleTypeDrawable) actual;
    assertThat(scaleTypeActual.getScaleType()).isEqualTo(ScalingUtils.ScaleType.FIT_CENTER);
    assertThat(scaleTypeActual.getFocusPoint()).isEqualTo(expectedFocusPoint);
  }

  @Test
  public void testBuildOverlayRes_whenUnset_thenReturnNull() {
    ImageOptions options = ImageOptions.create().build();

    Drawable overlayDrawable = mHierarcher.buildOverlayDrawable(mResources, options);

    assertThat(overlayDrawable).isNull();
  }

  @Test
  public void testBuildOverlayRes_whenSet_thenReturnDrawable() {
    ImageOptions options = ImageOptions.create().overlayRes(RES_ID).build();

    Drawable overlayDrawable = mHierarcher.buildOverlayDrawable(mResources, options);

    assertThat(overlayDrawable).isEqualTo(mDrawable);
  }

  @Test(expected = Resources.NotFoundException.class)
  public void testBuildOverlayDrawable_whenInvalidResId_thenThrowNotFoundException() {
    ImageOptions options = ImageOptions.create().overlayRes(INVALID_RES_ID).build();

    mHierarcher.buildOverlayDrawable(mResources, options);
  }
}

File: vito/core-java-impl/src/test/java/com/facebook/fresco/vito/core/impl/HierarcherImplTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.core.impl

import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.PointF
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.util.DisplayMetrics
import com.facebook.drawee.drawable.RoundedBitmapDrawable
import com.facebook.drawee.drawable.RoundedColorDrawable
import com.facebook.drawee.drawable.ScaleTypeDrawable
import com.facebook.drawee.drawable.ScalingUtils
import com.facebook.fresco.vito.options.ImageOptions
import com.facebook.fresco.vito.options.ImageOptionsDrawableFactory
import com.facebook.fresco.vito.options.RoundingOptions
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.fail
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.AdditionalMatchers
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class HierarcherImplTest {

  private lateinit var resources: Resources
  private lateinit var displayMetrics: DisplayMetrics
  private lateinit var drawable: Drawable
  private lateinit var hierarcher: Hierarcher
  private lateinit var drawableFactory: ImageOptionsDrawableFactory

  @Before
  fun setup() {
    resources = mock()
    drawable = mock()
    drawableFactory = mock()
    displayMetrics = DisplayMetrics()
    whenever(resources.displayMetrics).thenReturn(displayMetrics)
    whenever(resources.getDrawable(eq(RES_ID))).thenReturn(drawable)
    whenever(resources.getDrawable(AdditionalMatchers.not(eq(RES_ID))))
        .thenThrow(Resources.NotFoundException())
    hierarcher = HierarcherImpl(drawableFactory)
  }

  @Test
  fun testBuildPlaceholderRes() {
    val options = ImageOptions.create().placeholderRes(RES_ID).build()
    val errorDrawable = hierarcher.buildPlaceholderDrawable(resources, options)
    assertThat(errorDrawable).isNotNull()
    assertThat(errorDrawable).isInstanceOf(ScaleTypeDrawable::class.java)
    val scaleTypeDrawable = errorDrawable as ScaleTypeDrawable
    assertThat(scaleTypeDrawable.scaleType).isEqualTo(ImageOptions.defaults().placeholderScaleType)
    assertThat(scaleTypeDrawable.focusPoint)
        .isEqualTo(ImageOptions.defaults().placeholderFocusPoint)
    assertThat(scaleTypeDrawable.current).isEqualTo(drawable)
  }

  @Test
  fun testBuildPlaceholderDrawable() {
    val expected = ColorDrawable(Color.YELLOW)
    val options = ImageOptions.create().placeholder(expected).placeholderScaleType(null).build()
    val result = hierarcher.buildPlaceholderDrawable(resources, options)
    assertThat(result).isEqualTo(expected)
  }

  @Test
  fun testBuildPlaceholderDrawableScale() {
    val expected = ColorDrawable(Color.YELLOW)
    val options =
        ImageOptions.create()
            .placeholder(expected)
            .placeholderScaleType(ScalingUtils.ScaleType.CENTER)
            .build()
    val result = hierarcher.buildPlaceholderDrawable(resources, options)
    if (result == null) {
      fail("not null value expected")
      return
    }
    assertThat(result).isExactlyInstanceOf(ScaleTypeDrawable::class.java)
    assertThat(result.current).isEqualTo(expected)
  }

  @Test
  fun testApplyRoundingOptions_whenRoundAsCircle_thenReturnDrawable() {
    val drawable = ColorDrawable(Color.YELLOW)
    whenever(resources.getDrawable(RES_ID)).thenReturn(drawable)
    val options = mock<ImageOptions>()
    whenever(options.roundingOptions).thenReturn(RoundingOptions.asCircle())
    whenever(options.placeholderDrawable).thenReturn(drawable)
    whenever(options.placeholderApplyRoundingOptions).thenReturn(true)
    var result = hierarcher.buildPlaceholderDrawable(resources, options)
    assertThat(result).isExactlyInstanceOf(RoundedColorDrawable::class.java)
    whenever(options.placeholderDrawable).thenReturn(null)
    whenever(options.placeholderRes).thenReturn(RES_ID)
    result = hierarcher.buildPlaceholderDrawable(resources, options)
    assertThat(result).isExactlyInstanceOf(RoundedColorDrawable::class.java)
  }

  @Test
  fun testApplyRoundingOptions_whenRoundWithCornerRadius_thenReturnDrawable() {
    val drawable = mock<BitmapDrawable>()
    val bitmap = mock<Bitmap>()
    whenever(drawable.bitmap).thenReturn(bitmap)
    val options = mock<ImageOptions>()
    whenever(options.roundingOptions).thenReturn(RoundingOptions.forCornerRadiusPx(123f))
    whenever(options.placeholderDrawable).thenReturn(drawable)
    whenever(options.placeholderApplyRoundingOptions).thenReturn(true)
    val result = hierarcher.buildPlaceholderDrawable(resources, options)
    assertThat(result).isNotNull()
    assertThat(result).isInstanceOf(RoundedBitmapDrawable::class.java)
    val roundedBitmapDrawable = result as RoundedBitmapDrawable
    assertThat(roundedBitmapDrawable.radii)
        .isEqualTo(floatArrayOf(123f, 123f, 123f, 123f, 123f, 123f, 123f, 123f))
  }

  @Test
  fun testApplyRoundingOptions_whenRoundWithCornerRadii_thenReturnDrawable() {
    val drawable = mock<BitmapDrawable>()
    val bitmap = mock<Bitmap>()
    whenever(drawable.bitmap).thenReturn(bitmap)
    val options = mock<ImageOptions>()
    whenever(options.roundingOptions).thenReturn(RoundingOptions.forCornerRadii(1f, 2f, 3f, 4f))
    whenever(options.placeholderDrawable).thenReturn(drawable)
    whenever(options.placeholderApplyRoundingOptions).thenReturn(true)
    val result = hierarcher.buildPlaceholderDrawable(resources, options)
    assertThat(result).isNotNull()
    assertThat(result).isInstanceOf(RoundedBitmapDrawable::class.java)
    val roundedBitmapDrawable = result as RoundedBitmapDrawable
    assertThat(roundedBitmapDrawable.radii).isEqualTo(floatArrayOf(1f, 1f, 2f, 2f, 3f, 3f, 4f, 4f))
  }

  @Test
  fun testBuildErrorDrawable_whenNoScaleTypeSet_thenUseDefaultScaleType() {
    val options = ImageOptions.create().errorRes(RES_ID).build()
    val errorDrawable = hierarcher.buildErrorDrawable(resources, options)
    assertThat(errorDrawable).isNotNull()
    assertThat(errorDrawable).isInstanceOf(ScaleTypeDrawable::class.java)
    val scaleTypeDrawable = errorDrawable as ScaleTypeDrawable
    assertThat(scaleTypeDrawable.scaleType).isEqualTo(ImageOptions.defaults().errorScaleType)
    assertThat(scaleTypeDrawable.focusPoint).isEqualTo(ImageOptions.defaults().errorFocusPoint)
    assertThat(scaleTypeDrawable.current).isEqualTo(drawable)
  }

  @Test
  fun testBuildErrorDrawable_whenScaleTypeNull_thenDoNotWrapDrawable() {
    val options = ImageOptions.create().errorRes(RES_ID).errorScaleType(null).build()
    val errorDrawable = hierarcher.buildErrorDrawable(resources, options)
    assertThat(errorDrawable).isEqualTo(drawable)
  }

  @Test(expected = Resources.NotFoundException::class)
  fun testBuildErrorDrawable_whenInvalidResId_thenThrowNotFoundException() {
    val options = ImageOptions.create().errorRes(INVALID_RES_ID).build()
    hierarcher.buildErrorDrawable(resources, options)
  }

  @Test
  fun testBuildErrorDrawable_whenScaleTypeSet_thenReturnScaleTypeDrawable() {
    val focusPoint = PointF(100f, 234f)
    val options =
        ImageOptions.create()
            .errorRes(RES_ID)
            .errorScaleType(ScalingUtils.ScaleType.FOCUS_CROP)
            .errorFocusPoint(focusPoint)
            .build()
    val errorDrawable = hierarcher.buildErrorDrawable(resources, options)
    assertThat(errorDrawable).isNotNull()
    assertThat(errorDrawable).isInstanceOf(ScaleTypeDrawable::class.java)
    val scaleTypeDrawable = errorDrawable as ScaleTypeDrawable
    assertThat(scaleTypeDrawable.scaleType).isEqualTo(ScalingUtils.ScaleType.FOCUS_CROP)
    assertThat(scaleTypeDrawable.focusPoint).isEqualTo(focusPoint)
    assertThat(scaleTypeDrawable.current).isEqualTo(drawable)
  }

  @Test
  fun testBuildErrorDrawable_whenNotSet_thenReturnNopDrawable() {
    val options = ImageOptions.create().build()
    val errorDrawable = hierarcher.buildErrorDrawable(resources, options)
    assertThat(errorDrawable).isNull()
  }

  @Test
  fun testBuildProgressDrawable() {
    val drawable = ColorDrawable(0x0)
    val imageOptions =
        ImageOptions.create()
            .progress(drawable)
            .progressScaleType(ScalingUtils.ScaleType.FIT_CENTER)
            .build()
    val actual = hierarcher.buildProgressDrawable(resources, imageOptions)
    assertThat(actual).isInstanceOf(ScaleTypeDrawable::class.java)
    val scaleTypeActual = actual as ScaleTypeDrawable
    assertThat(scaleTypeActual.scaleType).isEqualTo(ScalingUtils.ScaleType.FIT_CENTER)
    assertThat(scaleTypeActual.current).isEqualTo(drawable)
  }

  @Test
  fun testBuildActualImageWrapper() {
    val expectedFocusPoint = PointF(1f, 2f)
    val imageOptions =
        ImageOptions.create()
            .scale(ScalingUtils.ScaleType.FIT_CENTER)
            .focusPoint(expectedFocusPoint)
            .build()
    val actual: Drawable = hierarcher.buildActualImageWrapper(imageOptions, null)
    assertThat(actual).isInstanceOf(ScaleTypeDrawable::class.java)
    val scaleTypeActual = actual as ScaleTypeDrawable
    assertThat(scaleTypeActual.scaleType).isEqualTo(ScalingUtils.ScaleType.FIT_CENTER)
    assertThat(scaleTypeActual.focusPoint).isEqualTo(expectedFocusPoint)
  }

  @Test
  fun testBuildOverlayRes_whenUnset_thenReturnNull() {
    val options = ImageOptions.create().build()
    val overlayDrawable = hierarcher.buildOverlayDrawable(resources, options)
    assertThat(overlayDrawable).isNull()
  }

  @Test
  fun testBuildOverlayRes_whenSet_thenReturnDrawable() {
    val options = ImageOptions.create().overlayRes(RES_ID).build()
    val overlayDrawable = hierarcher.buildOverlayDrawable(resources, options)
    assertThat(overlayDrawable).isEqualTo(drawable)
  }

  @Test(expected = Resources.NotFoundException::class)
  fun testBuildOverlayDrawable_whenInvalidResId_thenThrowNotFoundException() {
    val options = ImageOptions.create().overlayRes(INVALID_RES_ID).build()
    hierarcher.buildOverlayDrawable(resources, options)
  }

  companion object {
    private const val RES_ID = 123
    private const val INVALID_RES_ID = 999
  }
}
